ynput/OpenPype
# Clockify: refresh and fix the integration

## Changelog Description
Due to recent API changes, Clockify requires `user_id` to operate with the timers. 
I updated this part and currently it is a WIP for making it fully functional. Most functions, such as start and stop timer, and projects sync are currently working. For the rate limiting task new dependency is added: [https://pypi.org/project/ratelimiter/](url)

## Additional info
_TODO:_

- [x] ~~Fix the user permissions check. User validation is currently set to True for all users.~~
- [ ] Check for currently running timer, started with the App Launcher. The idea is to check for the currently running timer during application launch, if it is already running for the selected task. For some reason `get_in_progress` does not work in this case. This check works normally when the timer is started with `Clockify - Start Timer` action.
- [ ] Check if Clockify to Ftrack integration works. I currently have no Ftrack subscription, so I would be glad if someone could test it.

## Testing notes:
1. Enable Clockify module and enter workspace_name in the settings
2. Restart OP and fill the API key, generated in the Clockify Profile settings
3. Open Launcher and choose `Sync to Clockify`
4. Check that current OP projects now appear in the Clockify interface
5. Choose any OP task and choose `Clockify - Start Timer`
6. The timer will start, the timer description is in the format `asset_name/task_name`
7. Right click the OP tray and choose `Clockify` -> `Stop Timer`
8. Open Launcher and start any compatible application. 
9. The timer with the description set to the corresponding asset name will be started. If there's already running timer, it will be stopped. 
10. Launch OpenPype menu and switch to another workfile. The timer will change.
